### ISSUE TITLE
Error during build in docker image (for CPU and CUDA backends)

### ISSUE THREAD

**Issue by tlikhomanenko** [arrayfire/arrayfire#2767]
Hi, thanks for the new release!
Trying to build in docker image the new version 3.7.0 and have some troubles.

For CUDA backend I have error during cmake:
```
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0                                                                                                                                                                                                                                                                                                                             -- Check for working C compiler: /usr/bin/cc                                                                                                                                                                                                                                                                                                                                -- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - not found
-- Looking for CL_VERSION_1_2
-- Looking for CL_VERSION_1_2 - found
-- Could NOT find OpenCL (missing:  OpenCL_LIBRARY) (found version "1.2")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found CUDA: /usr/local/cuda (found suitable version "10.0", minimum required is "7.0")
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Found cuDNN: /usr/lib/x86_64-linux-gnu/libcudnn.so.7 (found suitable version "7.6", minimum required is "4.0")
-- Could NOT find OpenCL (missing:  OpenCL_LIBRARY) (found suitable version "1.2", minimum required is "1.2")
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libGL.so
-- Could NOT find FreeImage (missing:  FreeImage_INCLUDE_DIR FreeImage_LINK_LIBRARY)
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Could NOT find FFTW (missing:  FFTW_INCLUDE_DIR FFTW_LIBRARY FFTWF_LIBRARY)
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of void*
-- Check size of void* - done
-- Checking for [Accelerate]
-- Checking for [vecLib]
-- Checking for [cblas - atlas]
-- Checking for [openblas]
-- Checking for [cblas]
-- Checking for [blas]
-- CBLAS library not found.
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Could NOT find LAPACK (missing:  LAPACK_INCLUDE_DIR LAPACK_LIBRARIES)
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE)
-- Check size of int
-- Check size of int - done
-- MKL: Thread Layer(Intel OpenMP) Interface(4-byte Integer)
VERBOSEMKL environment variable(LIBRARY_PATH): /usr/local/cuda/lib64/stubs
VERBOSEMKL_ThreadingLibrary_LINK_LIBRARY: /usr/lib/x86_64-linux-gnu/libiomp5.so
-- Could NOT find MKL: Source the compilervars.sh or mklvars.sh scripts included with your installation of MKL. This script searches for the libraries in MKLROOT, LIBRARY_PATHS(Linux), and LIB(Windows) environment variables (missing:  MKL_INCLUDE_DIR MKL_Core_LINK_LIBRARY MKL_Interface_LINK_LIBRARY MKL_ThreadLayer_LINK_LIBRARY)
-- Could NOT find MKL: Source the compilervars.sh or mklvars.sh scripts included with your installation of MKL. This script searches for the libraries in MKLROOT, LIBRARY_PATHS(Linux), and LIB(Windows) environment variables (missing:  MKL_INCLUDE_DIR MKL_Core_STATIC_LINK_LIBRARY MKL_Interface_STATIC_LINK_LIBRARY MKL_ThreadLayer_STATIC_LINK_LIBRARY)
-- Could NOT find Boost
CMake Warning at CMakeModules/boost_package.cmake:36 (message):
  WARN: Found Boost v..  Required 1.70.0.  Build will download Boost Compute.
Call Stack (most recent call first):
  CMakeLists.txt:48 (include)


-- Automatic GPU detection failed. Building for common architectures.
-- CUDA_architecture_build_targets: 3.0;3.5;5.0;5.2;6.0;6.1;7.0;7.0+PTX
-- Performing Test has_ignored_attributes_flag
-- Performing Test has_ignored_attributes_flag - Success
-- CUDA driver library missing. Looking for libcuda stub.
-- CUDA driver stub FOUND: /usr/local/cuda/lib64/stubs/libcuda.so
CMake Error at /usr/share/cmake-3.5/Modules/ExternalProject.cmake:1888 (message):
  Could not extract tarball filename from url:

    https://sparse.tamu.edu/MM/JGD_Kocay/Trec4.tar.gz;DOWNLOAD_NO_EXTRACT;False
Call Stack (most recent call first):
  /usr/share/cmake-3.5/Modules/ExternalProject.cmake:2459 (_ep_add_download_command)
  test/CMakeModules/download_sparse_datasets.cmake:19 (ExternalProject_Add)
  test/CMakeModules/download_sparse_datasets.cmake:37 (mtxDownload)
  test/CMakeLists.txt:14 (include)


-- Configuring incomplete, errors occurred!
```

and for CPU backend:
```
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - not found
-- Looking for CL_VERSION_1_2
-- Looking for CL_VERSION_1_2 - not found
-- Looking for CL_VERSION_1_1
-- Looking for CL_VERSION_1_1 - not found
-- Looking for CL_VERSION_1_0
-- Looking for CL_VERSION_1_0 - not found
-- Could NOT find OpenCL (missing: OpenCL_LIBRARY OpenCL_INCLUDE_DIR)
CUDA_TOOLKIT_ROOT_DIR not found or specified
-- Could NOT find CUDA (missing: CUDA_TOOLKIT_ROOT_DIR CUDA_NVCC_EXECUTABLE CUDA_INCLUDE_DIRS CUDA_CUDART_LIBRARY) (Required is at least version "7.0")
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
-- Could NOT find cuDNN (missing: cuDNN_LINK_LIBRARY cuDNN_INCLUDE_DIRS) (Required is at least version "4.0")
-- Could NOT find OpenCL (missing: OpenCL_LIBRARY OpenCL_INCLUDE_DIR) (Required is at least version "1.2")
-- Could NOT find OpenGL (missing: OPENGL_gl_LIBRARY OPENGL_INCLUDE_DIR)
-- Could NOT find FreeImage (missing: FreeImage_INCLUDE_DIR FreeImage_LINK_LIBRARY)
-- Looking for pthread.h
-- Looking for pthread.h - found                                                                                                                                                                                                                                                                                                                                            -- Looking for pthread_create                                                                                                                                                                                                                                                                                                                                               -- Looking for pthread_create - not found                                                                                                                                                                                                                                                                                                                                   -- Looking for pthread_create in pthreads                                                                                                                                                                                                                                                                                                                                   -- Looking for pthread_create in pthreads - not found                                                                                                                                                                                                                                                                                                                       -- Looking for pthread_create in pthread                                                                                                                                                                                                                                                                                                                                    -- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
-- Found FFTW: /usr/include
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of void*
-- Check size of void* - done
-- Checking for [Accelerate]
-- Checking for [vecLib]
-- Checking for [cblas - atlas]
-- Includes found
-- Looking for cblas_dgemm
-- Looking for cblas_dgemm - found
-- CBLAS Symbols FOUND
-- CBLAS library found
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE)
-- Found LAPACK: /usr/include
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE)
-- Check size of int
-- Check size of int - done
-- MKL: Thread Layer(Intel OpenMP) Interface(4-byte Integer)
VERBOSEMKL environment variable(LIBRARY_PATH):
VERBOSEMKL_Core_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_core.so
VERBOSEMKL_Core_STATIC_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_core.a
VERBOSEMKL_RT_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_rt.so
VERBOSEMKL_ThreadLayer_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_intel_thread.so
VERBOSEMKL_ThreadLayer_STATIC_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_intel_thread.a
VERBOSEMKL_ThreadingLibrary_LINK_LIBRARY: /usr/lib/x86_64-linux-gnu/libiomp5.so
VERBOSEMKL_Interface_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_intel_lp64.so
VERBOSEMKL_Interface_STATIC_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_intel_lp64.a
VERBOSEMKL_mkl_def_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_def.so
VERBOSEMKL_mkl_mc_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_mc.so
VERBOSEMKL_mkl_mc3_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_mc3.so
VERBOSEMKL_mkl_avx_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_avx.so
VERBOSEMKL_mkl_avx2_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_avx2.so
VERBOSEMKL_mkl_avx512_LINK_LIBRARY: /opt/intel/mkl/lib/intel64/libmkl_avx512.so
-- Found MKL_Shared: /opt/intel/mkl/include (found version "2018.0.4")
-- Found MKL_Static: /opt/intel/mkl/include (found version "2018.0.4")
-- Could NOT find Boost
CMake Warning at CMakeModules/boost_package.cmake:36 (message):
  WARN: Found Boost v..  Required 1.70.0.  Build will download Boost Compute.
Call Stack (most recent call first):
  CMakeLists.txt:48 (include)


-- Performing Test has_ignored_attributes_flag
-- Performing Test has_ignored_attributes_flag - Success
-- Could NOT find PythonInterp (missing: PYTHON_EXECUTABLE)
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Configuring done
CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_uchar":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_uchar":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_char":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_char":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_double":

    "Boost_INCLUDE_DIR-NOTFOUND"
CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_double":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_uintl":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_uintl":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_short":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_short":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_float":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Found relative path while evaluating include directories of
  "cpu_sort_by_key_float":

    "Boost_INCLUDE_DIR-NOTFOUND"



CMake Error in src/backend/cpu/CMakeLists.txt:
  Target "c_api_interface" contains relative path in its
  INTERFACE_INCLUDE_DIRECTORIES:

    "Boost_INCLUDE_DIR-NOTFOUND"
....
```
in case I add `sudo apt-get install libboost-all-dev` for CPU backend the following error occurs during make:
```
/tmp/arrayfire/test/mmio/mmio.c: In function 'mm_read_unsymmetric_sparse':
/tmp/arrayfire/test/mmio/mmio.c:68:9: warning: ignoring return value of 'fscanf', declared with attribute warn_unused_result [-Wunused-result]
         fscanf(f, "%d %d %lg\n", &I[i], &J[i], &val[i]);
         ^
[  3%] Performing download step (download, verify and extract) for 'boost_compute'
[  4%] Performing download step (download and verify) for 'mtxDownload-LPnetlib-lpi_vol1'
[  4%] Performing download step (download and verify) for 'mtxDownload-HB-bcsstm02'
[  4%] Performing download step (download and verify) for 'mtxDownload-QCD-conf6_0-4x4-20'
[  4%] Performing download step (download and verify) for 'mtxDownload-Sandia-oscil_dcop_42'
[  4%] Performing download step (download and verify) for 'mtxDownload-Sandia-oscil_dcop_12'
-- Downloading...
   dst='/tmp/arrayfire/build/boost_compute-prefix/src/boost-1.70.0.tar.gz'
   timeout='none'
In file included from /tmp/arrayfire/src/backend/cpu/./err_cpu.hpp:10:0,
                 from /tmp/arrayfire/src/backend/cpu/./kernel/sort_by_key_impl.hpp:12,
                 from /tmp/arrayfire/src/backend/cpu/kernel/sort_by_key/sort_by_key_impl.cpp:10:
/tmp/arrayfire/src/backend/cpu/../common/err_common.hpp:12:32: fatal error: boost/stacktrace.hpp: No such file or directory
compilation terminated.
-- Using src='https://github.com/boostorg/compute/archive/boost-1.70.0.tar.gz'
[  4%] Performing download step (download and verify) for 'mtxDownload-QCD-conf6_0-4x4-30'
In file included from /tmp/arrayfire/src/backend/cpu/./err_cpu.hpp:10:0,
                 from /tmp/arrayfire/src/backend/cpu/./kernel/sort_by_key_impl.hpp:12,                                                                                                                                       from /tmp/arrayfire/src/backend/cpu/kernel/sort_by_key/sort_by_key_impl.cpp:10:                                                                                                            /tmp/arrayfire/src/backend/cpu/../common/err_common.hpp:12:32: fatal error: boost/stacktrace.hpp: No such file or directory
compilation terminated.
```

I am using the following docker files:
- for CUDA:
```
FROM nvidia/cuda:10.0-cudnn7-devel-ubuntu16.04

RUN APT_INSTALL="apt-get install -y --no-install-recommends" && \
    rm -rf /var/lib/apt/lists/* \
           /etc/apt/sources.list.d/cuda.list \
           /etc/apt/sources.list.d/nvidia-ml.list && \
    apt-get update && \
    DEBIAN_FRONTEND=noninteractive $APT_INSTALL \
        build-essential \
        ca-certificates \
        cmake \
        wget \
        git \
        vim \
        emacs \
        nano \
        htop \
        g++ \
        # ssh for OpenMPI
        openssh-server openssh-client \
        # OpenMPI
        libopenmpi-dev libomp-dev \
        # nccl: for flashlight
        libnccl2 libnccl-dev \
        libglfw3-dev && \
# ==================================================================
# arrayfire https://github.com/arrayfire/arrayfire/wiki/
# ------------------------------------------------------------------
    cd /tmp && git clone --recursive https://github.com/arrayfire/arrayfire.git && \
    cd arrayfire && git checkout v3.7.0 && git submodule update --init --recursive && \
    mkdir build && cd build && \
    CXXFLAGS=-DOS_LNX cmake .. -DCMAKE_BUILD_TYPE=Release -DAF_BUILD_CPU=OFF -DAF_BUILD_OPENCL=OFF -DAF_BUILD_EXAMPLES=OFF && \
    make -j$(nproc) && \
    make install && \
# ==================================================================
# config & cleanup
# ------------------------------------------------------------------
    ldconfig && \
    apt-get clean && \
    apt-get autoremove && \
    rm -rf /var/lib/apt/lists/* /tmp/* ~/* && \
    # If the driver is not found (during docker build) the cuda driver api need to be linked against the
    # libcuda.so stub located in the lib[64]/stubs directory
    ln -s /usr/local/cuda/lib64/stubs/libcuda.so /usr/lib/x86_64-linux-gnu/libcuda.so.1
```
- for CPU:
```
FROM ubuntu:16.04

RUN APT_INSTALL="apt-get install -y --no-install-recommends" && \
    apt-get update && \
    DEBIAN_FRONTEND=noninteractive $APT_INSTALL \
        build-essential \
        ca-certificates \
        wget \
        git \
        vim \
        emacs \
        nano \
        htop \
        g++ \
        # for MKL
        apt-transport-https \
        # for arrayfire CPU backend
        # OpenBLAS
        libopenblas-dev libfftw3-dev liblapacke-dev \
        # ATLAS
        libatlas3-base libatlas-dev libfftw3-dev liblapacke-dev \
        # ssh for OpenMPI
        openssh-server openssh-client \
        # OpenMPI
        libopenmpi-dev libomp-dev && \
# ==================================================================
# cmake 3.10 (for MKLDNN)
# ------------------------------------------------------------------
    apt-get purge -y cmake && \
    # for cmake
    DEBIAN_FRONTEND=noninteractive $APT_INSTALL zlib1g-dev libcurl4-openssl-dev && \
    cd /tmp && wget https://cmake.org/files/v3.10/cmake-3.10.3.tar.gz  && \
    tar -xzvf cmake-3.10.3.tar.gz  && cd cmake-3.10.3  && \
    ./bootstrap --system-curl && \
    make -j$(nproc) &&  make install && cmake --version && \
# ==================================================================
# MKL https://software.intel.com/en-us/mkl
# ------------------------------------------------------------------
    cd /tmp && wget https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2019.PUB && \
    apt-key add GPG-PUB-KEY-INTEL-SW-PRODUCTS-2019.PUB && \
    wget https://apt.repos.intel.com/setup/intelproducts.list -O /etc/apt/sources.list.d/intelproducts.list && \
    sh -c 'echo deb https://apt.repos.intel.com/mkl all main > /etc/apt/sources.list.d/intel-mkl.list' && \
    apt-get update && DEBIAN_FRONTEND=noninteractive $APT_INSTALL intel-mkl-64bit-2018.4-057 && \
    export MKLROOT=/opt/intel/mkl && \
# ==================================================================
# arrayfire with CPU backend https://github.com/arrayfire/arrayfire/wiki/
# ------------------------------------------------------------------
    cd /tmp && git clone --recursive https://github.com/arrayfire/arrayfire.git && \
    cd arrayfire && git checkout v3.7.0  && git submodule update --init --recursive && \
    mkdir build && cd build && \
    CXXFLAGS=-DOS_LNX cmake .. -DCMAKE_BUILD_TYPE=Release -DAF_BUILD_CUDA=OFF -DAF_BUILD_OPENCL=OFF -DAF_BUILD_EXAMPLES=OFF && \
    make -j$(nproc) && make install && \
# ==================================================================
# MKLDNN https://github.com/intel/mkl-dnn/
# ------------------------------------------------------------------
    cd /tmp && git clone https://github.com/intel/mkl-dnn.git && \
    cd mkl-dnn && git checkout 4bdffc2 && mkdir -p build && cd build && \
    cmake .. -DMKLDNN_USE_MKL=FULL && \
    make -j$(nproc) && make install && \
# ==================================================================
# Gloo https://github.com/facebookincubator/gloo.git
# ------------------------------------------------------------------
    cd /tmp && git clone https://github.com/facebookincubator/gloo.git && \
    cd gloo && git checkout b7e0906 && mkdir build && cd build && \
    cmake .. -DUSE_MPI=ON && \
    make -j$(nproc) && make install && \
# ==================================================================
# config & cleanup
# ------------------------------------------------------------------
    ldconfig && \
    apt-get clean && \
    apt-get autoremove && \
    rm -rf /var/lib/apt/lists/* /tmp/* ~/*
```

This docker files worked fine with 3.6.4 version.

Thanks for the help!

**Comment by umar456** [comment]
What version of boost do you have on the system? We require boost stacktrace for this version.

**Comment by umar456** [comment]
We install boost using the following commands:

```
# Boost
RUN cd /usr/src && \
    wget https://dl.bintray.com/boostorg/release/1.70.0/source/boost_1_70_0.tar.gz && \
    tar xf boost_1_70_0.tar.gz && \
    mv boost_1_70_0 /artifact && \
    rm -rf boost_1_70_0*

```

**Comment by tlikhomanenko** [comment]
> What version of boost do you have on the system? We require boost stacktrace for this version.

I am using image of clear ubuntu 16.04, so as I know there is no boost. And previously I didn't need to install it. Did you add this additional requirement for both CPU and CUDA backend in 3.7.0 (don't see this requirement on the wiki https://github.com/arrayfire/arrayfire/wiki/Build-Instructions-for-Linux)? 

Will try to add installation of boost 1.70.0 in my docker image and will come back.

**Comment by umar456** [comment]
Sorry about that. I have updated the instructions. Thanks for pointing that out.

**Comment by 9prady9** [comment]
@tlikhomanenko If the issue has been resolved, please close the issue. Thank you.

**Comment by tlikhomanenko** [comment]
Hi @umar456, @9prady9 

I added 

```
RUN cd /tmp && wget https://dl.bintray.com/boostorg/release/1.70.0/source/boost_1_70_0.tar.gz && \
    tar xf boost_1_70_0.tar.gz && rm -rf boost_1_70_0.tar.gz && \
    cp -r /tmp/boost_1_70_0/boost /usr/local/include && \
    cd /tmp && git clone --recursive https://github.com/arrayfire/arrayfire.git && \
    cd arrayfire && git checkout v3.7.0 && git submodule update --init --recursive && \
    mkdir build && cd build && \
    CXXFLAGS=-DOS_LNX cmake .. -DCMAKE_BUILD_TYPE=Release -DAF_BUILD_CPU=OFF -DAF_BUILD_OPENCL=OFF -DAF_BUILD_EXAMPLES=OFF && \
    make -j$(nproc) && \
    make install && \
```

So now boost is found during cmake for CUDA backend, but this error still persists `Could not extract tarball filename from url:`

Full log of cmake:
```
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - not found
-- Looking for CL_VERSION_1_2
-- Looking for CL_VERSION_1_2 - found
-- Could NOT find OpenCL (missing:  OpenCL_LIBRARY) (found version "1.2")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found CUDA: /usr/local/cuda (found suitable version "10.0", minimum required is "7.0")
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Found cuDNN: /usr/lib/x86_64-linux-gnu/libcudnn.so.7 (found suitable version "7.6", minimum required is "4.0")
-- Could NOT find OpenCL (missing:  OpenCL_LIBRARY) (found suitable version "1.2", minimum required is "1.2")
-- Found OpenGL: /usr/lib/x86_64-linux-gnu/libGL.so
-- Could NOT find FreeImage (missing:  FreeImage_INCLUDE_DIR FreeImage_LINK_LIBRARY)
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Could NOT find FFTW (missing:  FFTW_INCLUDE_DIR FFTW_LIBRARY FFTWF_LIBRARY)
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of void*
-- Check size of void* - done
-- Checking for [Accelerate]
-- Checking for [vecLib]
-- Checking for [cblas - atlas]
-- Checking for [openblas]
-- Checking for [cblas]
-- Checking for [blas]
-- CBLAS library not found.
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Could NOT find LAPACK (missing:  LAPACK_INCLUDE_DIR LAPACK_LIBRARIES)
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE)
-- Check size of int
-- Check size of int - done
-- MKL: Thread Layer(Intel OpenMP) Interface(4-byte Integer)
VERBOSEMKL environment variable(LIBRARY_PATH): /usr/local/cuda/lib64/stubs
VERBOSEMKL_ThreadingLibrary_LINK_LIBRARY: /usr/lib/x86_64-linux-gnu/libiomp5.so
-- Could NOT find MKL: Source the compilervars.sh or mklvars.sh scripts included with your installation of MKL. This script searches for the libraries in MKLROOT, LIBRARY_PATHS(Linux), and LIB(Windows) environment variables (missing:  MKL_INCLUDE_DIR MKL_Core_LINK_LIBRARY MKL_Interface_LINK_LIBRARY MKL_ThreadLayer_LINK_LIBRARY)
-- Could NOT find MKL: Source the compilervars.sh or mklvars.sh scripts included with your installation of MKL. This script searches for the libraries in MKLROOT, LIBRARY_PATHS(Linux), and LIB(Windows) environment variables (missing:  MKL_INCLUDE_DIR MKL_Core_STATIC_LINK_LIBRARY MKL_Interface_STATIC_LINK_LIBRARY MKL_ThreadLayer_STATIC_LINK_LIBRARY)
-- Boost version: 1.70.0
-- Automatic GPU detection failed. Building for common architectures.
-- CUDA_architecture_build_targets: 3.0;3.5;5.0;5.2;6.0;6.1;7.0;7.0+PTX
-- Performing Test has_ignored_attributes_flag
-- Performing Test has_ignored_attributes_flag - Success
-- CUDA driver library missing. Looking for libcuda stub.
-- CUDA driver stub FOUND: /usr/local/cuda/lib64/stubs/libcuda.so
CMake Error at /usr/share/cmake-3.5/Modules/ExternalProject.cmake:1888 (message):
  Could not extract tarball filename from url:

    https://sparse.tamu.edu/MM/JGD_Kocay/Trec4.tar.gz;DOWNLOAD_NO_EXTRACT;False
Call Stack (most recent call first):
  /usr/share/cmake-3.5/Modules/ExternalProject.cmake:2459 (_ep_add_download_command)
  test/CMakeModules/download_sparse_datasets.cmake:19 (ExternalProject_Add)
  test/CMakeModules/download_sparse_datasets.cmake:37 (mtxDownload)
  test/CMakeLists.txt:14 (include)


-- Configuring incomplete, errors occurred!
```

**Comment by umar456** [comment]
Can you pass  -DAF_TEST_WITH_MTX_FILES=NO to your cmake command. It should avoid downloading those files. They are only there for sparse tests.

**Comment by 9prady9** [comment]
You can turn off mtx data based tests by turning OFF the following cmake variable.

AF_TEST_WITH_MTX_FILES

That should do the trick as work around. I will try to check if i can reproduce this on docker.

**Comment by 9prady9** [comment]
Sorry about the issue. I know what happened, the download no extract option for external project cmake command is not available in cmake 3.5.2. I will send in a fix soon. Thanks!

**Comment by tlikhomanenko** [comment]
Ohh, I see. Thanks! Will use this flag `-DAF_TEST_WITH_MTX_FILES=NO`

**Comment by umar456** [comment]
I see you are using different versions of cmake on the different backends. It seems we haven't tested in the older CMake which does not create the Booost::boost target. I think we need to update our ci to take this into account or update the minimal versions. For now I suggest you use the same command you use to install CMake on the cpu backend in the CUDA backend for now.

**Comment by tlikhomanenko** [comment]
Yep, the error occurs because of old cmake version (tested right before you @umar456 posted). Using now 3.10 for both CPU and CUDA backends and everything is configured correctly.

Could you add in the build instructions the cmake version requirements (the one you are using, so no troubles with this anymore :) )? 

During build for CUDA backend now the error during make 
```
In file included from /tmp/arrayfire/src/backend/cuda/cudnnModule.cpp:13:0:
/tmp/arrayfire/src/backend/cuda/cudnnModule.hpp: In member function 'std::tuple<int, int, int> cuda::cudnnModule::getVersion()':
/tmp/arrayfire/src/backend/cuda/cudnnModule.hpp:71:38: error: converting to 'std::tuple<int, int, int>' from initializer list would use explicit constructor 'constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {int&, int&, int&}; <template-parameter-2-2> = void; _Elements = {int, int, int}]'
         return { major, minor, patch };
                                      ^
```

If I switch to Ubuntu 18.04 docker image make works fine. So seems maybe the problem in gcc version.

Switching to Ubuntu 18.04 in Dockerfiles. Can confirm that build is working for CPU and CUDA backend.

**Comment by 9prady9** [comment]
@tlikhomanenko With fixes from PRs #2771 and #2772 . You should be able to just the cmake(3.5.1) package from Ubuntu 16.04 package manager. You would however need version of boost that has stacktrace though.

**Comment by tlikhomanenko** [comment]
If I am using cmake 3.13 then I don't need to add any boost installation steps, it will be done automatically during cmake.
